js/editor-ta.js

```diff
@@ -0,0 +1,304 @@
+}
+
+/**
+ * @param {string} opt_content
+ * @return {EditSession}
+ * Create an edit session for a new file. Each tab should have its own session.
+ */
+EditorTextArea.prototype.newSession = function(opt_content) {
+  // TODO(zafzal): this
+  // Note opt_content is inital content to load into the area, remember to
+  // update the this.numLines_ and stuff;
+  // maybe just called updateLineNumbers and updateTextArea}
+};
+
+/**
+ * @param {EditSession} session
+ * Change the current session, usually to switch to another tab.
+ */
+EditorTextArea.prototype.setSession = function(session) {
+  // TODO(zafzal): this
+  // Note when switching out content remember to update the this.numLines_ and stuff;
+  // maybe just called updateLineNumbers and updateTextArea
+};
+
+/**
+ * @return {Search}
+ * Return search object.
+ */
+EditorTextArea.prototype.getSearch = function() {
+  // TODO(zafzal): this
+};
+
+EditorTextArea.prototype.onChange = function() {
+  // TODO(zafzal): Speed this up
+  this.numLines_ = this.textarea_.value.split('').filter(x => x === '\n').length + 1;
+  this.mirror_.innerText = this.textarea_.value;
+
+  // If we have a blank line at the end we have to add a character to get the
+  // height to set correctly
+  if (this.textarea_.value[this.textarea_.value.length-1] === '\n') {
+    this.mirror_.innerText += 'A';
+  }
+  this.updateLineNumbers();
+  this.updateTextArea();
+  // TODO(zafzal): need to emit a docChange event to each tab can know a change occured...
+};
+
+EditorTextArea.prototype.updateTextArea = function() {
+  const calculatedHeight = this.mirror_.getBoundingClientRect().height;
+  const calculatedWidth = this.mirror_.getBoundingClientRect().width;
+  // if the height is less then the window height just snap to the window height
+  if (calculatedHeight < this.initHeight_) {
+    this.container_.scrollTop = 0;
+    this.updateHeight(this.initHeight_);
+  } else {
+    this.updateHeight(calculatedHeight);
+  }
+
+  if (this.settings_.get('wraplines')) {
+    this.updateWidth(this.initWidth_);
+  } else {
+    this.updateWidth(Math.max(this.initWidth_, calculatedWidth));
+  }
+
+};
+
+
+EditorTextArea.prototype.updateLineNumbers = function() {
+  let linesRendered = this.lineNumbers_.children.length;
+
+  // remove excess elements
+  while (linesRendered > this.numLines_) {
+    this.lineNumbers_.lastChild.remove();
+    linesRendered--;
+  }
+  // add missing elements
+  while(linesRendered < this.numLines_) {
+    const line = linesRendered + 1;
+    // If word wrap is on we have to additionally set height
+    const height = this.settings_.get('wordwrap') ? this.getLineHeight(line) : undefined;
+    this.lineNumbers_.appendChild(this.lineNumberElem(line, height));
+    linesRendered++;
+  }
+
+};
+
+EditorTextArea.prototype.getLineHeight = function(lineNumber) {
+  // TODO(zafzal): Speed this up, splitting the whole doc is a bottleneck
+  const line = this.textarea_.value.split('\n')[lineNumber-1];
```

Comment:
the lines aren't all the same height?

Reply:
nope! if you have word wrap on the height of the line changes depending on conten :(